### ISSUE TITLE
How to deal with license compliance?

### ISSUE THREAD

**Issue by AndersonTorres** [NixOS/nixpkgs#342570]
## Issue description

With the recent advent of Duckstation changing its license, some issues were raised about how Nixpkgs should deal with licensing compliance.

## References

Duckstation licensing issue: https://github.com/NixOS/nixpkgs/issues/341915

**Comment by Atemu** [comment]
I see two low hanging fruits:

1. Install license files into the package output. This is a bare minimum for license compliance and trivial to do. If the package script doesn't already do this by default (many do), it should be done manually. 
2. Use "branding" configureFlags where available. Many packages offer configure flags that set a branding string in the resulting program, signifying that it was packaged by us. This ought to be enough to comply with the license requirement because this branding implies that we modified it.

Beyond these, we should look into manually marking packages that do not offer such flags by i.e. inserting a `© Nixpkgs contributors` into the license text which the program displays to users. (The GPL requires interactive programs to do this. If it doesn't, I think we can assume the author doesn't care about following the GPL extremely strictly.)  
This is quite a bit more work, so I think the best course of action is to only do this when prompted to (i.e. upstream requests us to do it) or if it is conceivable that it might upset upstream if we don't. Most upstreams are pretty chill about this sort of thing but some aren't and it's those cases I think we should focus on. Duckstation is a good example.

**Comment by AndersonTorres** [comment]
In case of 1, I still prefer a separate output for this, at least by the typical rationale we provide for splitting outputs in the Nixpkgs manual.

Further, I have some conceptual concerns. Maybe it is too much perfectionism from my side, but oh well.

First, license metadata should be provided for all inputs of an expression - sources, patches, our own machinery etc.
Second, we should have something like a mathematical function that receives all the input above and output the appropriate course of action - things like "GPL requires interactive programs to display some info; create a compliance wrapper that accepts a --print-meta flag".

Third:
According to the manual, `meta` and `passthru` can't rebuild the expression.
However, some programs require further action according to their licensing info.

**Comment by emilazy** [comment]
I’m unconvinced that a separate output is sufficient for compliance, or that making every package multi‐output would be well‐received. In fact I suspect that we may be obligated to include the licence and copyright notice in *all* outputs.

We also cannot just copy template licence files, because the copyright notice differs between packages and must also be retained.

IIRC the interactive program thing doesn’t matter unless we apply patches, thankfully, and GPLv3 loosens the requirement further.

**Comment by emilazy** [comment]
```
  2. You may modify your copy or copies of the Program or any portion
of it, thus forming a work based on the Program, and copy and
distribute such modifications or work under the terms of Section 1
above, provided that you also meet all of these conditions:

[…]

    c) If the modified program normally reads commands interactively
    when run, you must cause it, when started running for such
    interactive use in the most ordinary way, to print or display an
    announcement including an appropriate copyright notice and a
    notice that there is no warranty (or else, saying that you provide
    a warranty) and that users may redistribute the program under
    these conditions, and telling the user how to view a copy of this
    License.  (Exception: if the Program itself is interactive but
    does not normally print such an announcement, your work based on
    the Program is not required to print an announcement.)
```

I was wrong; even the GPLv2 has the exception that means we don’t really have to care about it.

**Comment by emilazy** [comment]
BTW, I think that encoding full SPDX expressions in `meta.license` is probably a prerequisite to doing anything at all fancy here beyond my stdenv suggestion.

**Comment by Pandapip1** [comment]
How about we send the [machine-readable copyright information](https://dep-team.pages.debian.net/deps/dep5/) licenses information to `$outputDoc/copyright`?

**Comment by nyabinary** [comment]
Yeah, #351902 isn't under the right license right now, Pixar uses a modified Apache 2.0 license and currently the package is under the unmodified Apache 2.0 license.

**Comment by Atemu** [comment]
Please keep it on topic, this issue is not about wrong licenses in `meta`. There's nothing to discuss there, it just needs to be fixed.

**Comment by AndersonTorres** [comment]
> to `$outputDoc/copyright`, and have a `common-licenses` packages that can be symlinked to?

A specific `$outputLicense/copyright` is better, since it can be injected without modifying currently installed packages.

**Comment by emilazy** [comment]
I do not think a separate output is okay, since derivative works must include at least a pointer to the licence text and in many cases must include the licence text verbatim, and the whole point of outputs is that they are distributed separately and can be downloaded in isolation. The best option is probably for multiple‐output packages to grow a `license` output that is symlinked into every other output. Hopefully we could share licence text between packages, but distinct copyright notices will have to be preserved. There are probably SPDX standards that we could profitably use for this.

**Comment by Pandapip1** [comment]
> There are probably SPDX standards that we could profitably use for this.

That would be [REUSE](https://reuse.software/spec/).

**Comment by emilazy** [comment]
REUSE is for annotating source code upstream for licence and copyright information. REUSE‐annotated code would be a lot easier for us to automate compliance with, but is a rarity in the wild; we can’t rely on it in general, though it may make sense to build our tooling around it as an ideal case and then figure out how to extend it to the common case. The REUSE tool can automate producing an SPDX‐format SBOM document, which may be closer to the kind of thing we’d want to install with packages. There is also CycloneDX; frankly, I have no idea how they compare, and asking Google for comparisons results in a bunch of incomprehensible LLM sludge.

**Comment by Pandapip1** [comment]
Proposition: we start with REUSE compliance, and then ask package maintainers to upstream REUSE annotations.

**Comment by emilazy** [comment]
Full REUSE compliance is difficult to do retroactively and most projects won’t accept it as it adds a significant amount of annotation to every file in the repository. That’s not going to be an adequate approach to meeting our legal obligations under FOSS licences. It’s the entire NixOS project that’s exposed, not just individual package maintainers.

**Comment by Pandapip1** [comment]
I've made https://github.com/NixOS/nixpkgs/pull/353697, which is a naive setup hook that copies maintainer-specified license files into the default output. A review of it would be appreciated.

I'll continue looking into making a fully automated `reuse-license-compliance-hook`.

**Comment by fgaz** [comment]
> it adds a significant amount of annotation to every file in the repository

Not necessarily: https://reuse.software/faq/#bulk-license

**Comment by emilazy** [comment]
Yeah, although they make it clear in the FAQ and the specification that they’d really rather you use file headers.

**Comment by AndersonTorres** [comment]
I am cleaning up my backlogs, and closing things I no longer want to pursue.

Feel free to open under a new issue.

**Comment by Pandapip1** [comment]
Reopening as per https://github.com/NixOS/nixpkgs/issues/208242#issuecomment-3173014292

**Comment by LuNeder** [comment]
> Duckstation is now actively blocking building on NixOS and seems to have added an additional restriction for packaging on their license after this commit [stenzek/duckstation@b20c2b2](https://github.com/stenzek/duckstation/commit/b20c2b281f7bb3bb2ac0f7cf2d12a3668f7351e9)

To my best knowledge this has zero legal value if you're not directly distributing their files or editing them. Nothing can stop you from making a software completely separate from theirs (aka nix) that happens to download their software directly from them and interact (install/patch) with it.

Even if you didn't want to patch the nixos block, you can just unset the relevant environment variables before you build it and maybe figure out a way to provide a fake /etc/os-release.

That said, even if I like the idea of making assholes angry, I personally believe people like this and their software do not deserve your time at all...

**Comment by Notarin** [comment]
Also yeah, reading some of these commits, this person is rather actively hostile, they are removing linux support entirely because apparently bug reports have infuriated them to that point.
This is not worth writing a derivation for just for it to vanish in a few days. Nor honestly should you want to use a piece of software with a developer who clearly does not like you.

My personal guess is they are a primarily windows user, and are frustrated by the fact Linux is different, and in their frustration are making a bit of a show and slandering the community a bit, and pulling support. Which, fine, sure, they're probably just frustrated. Not cool, but humans are humans.

My advice is to just get ahead of the blow and headcanon it as windows software and move on.

**Comment by Eveeifyeve** [comment]
> Correct me if I am wrong, but appimages still require patching. Which is by **far** illegal in this instance.

We are not lawyers at the end of that day, so we are not too sure if this is legal or not. So please just stop providing advice if your not qualified, your just making the situation complex and argumentive without being sure if it's right or wrong.

**Comment by PaulGrandperrin** [comment]
@Notarin 
IANAL but I'm pretty sure that  locally patching a binary:
- can only be illegal when bypassing a license check, encryption, or digital rights management, nothing else.
- can be a breach of contract when explicitly forbidden by a EULA coming with the binary. This can usually lead to things like licence or account termination.
- for the purpose of interoperability is actually protected by law in most jurisdictions (US and EU) 

And more importantly, the idea is not to modify and redistribute duckstation ourselves, but to publish a recipe that does it on the end user's computer automatically.

EDIT:
See section (f) of https://www.law.cornell.edu/uscode/text/17/1201

**Comment by VeilSilence** [comment]
> > Quick download, then `appimage-run`, and indeed, duckstation just crashed at game start.
> 
> [@VeilSilence](https://github.com/VeilSilence) [Try this](https://nixos.org/manual/nixos/unstable/#sec-custom-packages-prebuilt)

Alas, same error.
Tried to package like this:
<details><summary>Details</summary>
<p>

```
{
  appimageTools,
  fetchurl,
  pkgs,
}: let
  pname = "duckstation";
  version = "0.1-10693";

  src = fetchurl {
    url = "https://github.com/stenzek/duckstation/releases/download/v${version}/DuckStation-x64-SSE2.AppImage";
    hash = "sha256-ekFdfUTNt7PTry+lgj/UDu+GEO2CzBJqXBMyWOrGcAg=";
  };
in
  appimageTools.wrapType2 {
    inherit pname version src;

    extraPkgs = pkgs:
      with pkgs; [
        vulkan-loader
        libxcb-cursor
      ];
  }
```

</p>
</details> 

Now I have this error: 
```
/store/kavwy4v9p1jhf952dzipkjj5l1hjx110-qtbase-6.10.1/lib/qt-6/plugins/platforms/../../../libQt6WaylandClient.so.6: undefined symbol: _ZTI28QBasicPlatformVulkanInstance, version Qt_6_PRIVATE_API"
```
So, from my limited knowledge and understanding, patching is required indeed..

**Comment by PaulGrandperrin** [comment]
@Notarin In the end, the issues are mostly on the side of duckstation's author because as far as I can see, there's no legal way to forbid a user from patching a program to make it compatible with his computer.

The law is really explicitly on our side in this specific use case.

**Comment by Notarin** [comment]
> [@Notarin](https://github.com/Notarin) IANAL but I'm pretty sure that locally patching a binary:
> 
>     * can only be illegal when bypassing a license check, encryption, or digital rights management, nothing else.
> 
>     * can be a breach of contract when explicitly forbidden by a EULA coming with the binary. This can usually lead to things like licence or account termination.
> 
>     * for the purpose of interoperability is actually protected by law in most jurisdictions (US and EU)
> 
> 
> And more importantly, the idea is not to modify and redistribute duckstation ourselves, but to publish a recipe that does it on the end user's computer automatically.
> 
> EDIT: See section (f) of https://www.law.cornell.edu/uscode/text/17/1201

That is correct. I meant patching serverside. I originally missed the "unfree" bit. But yeah, this is mostly what I was saying [here](https://github.com/NixOS/nixpkgs/issues/342570#issuecomment-3829585726).
Publishing "instructions" is very strictly not distribution of the software, and policing such a thing is not possible, or even remotely reasonable, and if it was, it would be incredibly dystopic.

But, I do remain on the stance there is no point in supporting software that will cease upstream support any second. It may even happen *today*.
I just don't see a point in updating this in nixpkgs when there are alternatives and upstream is cutting us off any second now. imo another repo is a far better place for that.
Also it is usually a good idea to respect an upstream maintainers wishes, even if they are an arse about it, and they prefer the source of their software to be their web portal.

**Comment by SuperSamus** [comment]
> Quick download, then `appimage-run`, and indeed, duckstation just crashed at game start.

Works out of the box on my end...
Maybe because I have `programs.nix-ld.enable = true`?

The error message points to issues with Vulkan, so _maybe_ this solves it? (Throwing the idea randomly.)
```
export LD_LIBRARY_PATH=$(nix-build -E '(import <nixpkgs> {}).vulkan-loader' --no-out-link)/lib
```
Otherwise, maybe try the OpenGL rendering backend?

**Comment by zierf** [comment]
@VeilSilence The AppImage most likely bundles an incompatible QT version which crashes on KDE systems, I guess the reports for a working bundle are probably for non KDE systems.

<details>
<summary>Standard Crash Log</summary>

`appimage-run ./DuckStation-x64.AppImage`

```SH
XDG_SESSION_DESKTOP=KDE
Enabling xdg-desktop-portal platform theme.
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "" even though it was found.
qt.qpa.plugin: From 6.5.0, xcb-cursor0 or libxcb-cursor0 is needed to load the Qt xcb platform plugin.
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
qt.qpa.plugin: Could not find the Qt platform plugin "xcb,wayland" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, wayland-brcm, wayland-egl, wayland, xcb.

*************** Unhandled SIGABRT at 0x7fe48269df1c ***************
  0x00562157d54588 LogCallstack [../build/../src/common/crash_handler.cpp:389]
  0x00562157d54448 CrashSignalHandler [../build/../src/common/crash_handler.cpp:423]
  0x007fe482641d9f
  0x007fe48269df1c
  0x007fe482641ced
  0x007fe482628940
  0x007fe482ebeda5
  0x007fe482eff7c3
  0x007fe482ec0636
  0x007fe483b277f2
  0x007fe483bd57a6
  0x007fe483bd6217
  0x007fe482f8b3f0
  0x007fe483bd9214
  0x007fe484f985cc
  0x00562157a44902 main [../build/../src/duckstation-qt/qthost.cpp:3573]
  0x007fe48262a4c4
  0x007fe48262a577
  0x005621578c4754
  0xffffffffffffffff
*******************************************************************
Aborting application.
```
</details>

Enabling QT plugin debug messages reveals a problem with `Qt_6_PRIVATE_API`.

<details>
<summary>Extended Crash Log</summary>

`QT_DEBUG_PLUGINS=1 appimage-run ./DuckStation-x64.AppImage`

```SH
qt.core.plugin.factoryloader: Got keys from plugin meta data QList("xcb")
qt.core.plugin.factoryloader: checking directory path "~/.cache/appimage-run/5fe76f202d72dc2f2c2a7071fa91918b15275e9a8008e335323d930d3076b132/usr/bin/platforms" ...
qt.core.library: "/nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqwayland.so" cannot load: Cannot load library /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqwayland.so: /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/../../../libQt6WaylandClient.so.6: undefined symbol: _ZTI28QBasicPlatformVulkanInstance, version Qt_6_PRIVATE_API
qt.core.plugin.loader: QLibraryPrivate::loadPlugin failed on "/nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqwayland.so" : "Cannot load library /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqwayland.so: /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/../../../libQt6WaylandClient.so.6: undefined symbol: _ZTI28QBasicPlatformVulkanInstance, version Qt_6_PRIVATE_API"
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "" even though it was found.
qt.core.library: "/nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqxcb.so" cannot load: Cannot load library /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqxcb.so: /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/../../../libQt6XcbQpa.so.6: undefined symbol: _ZTI28QBasicPlatformVulkanInstance, version Qt_6_PRIVATE_API
qt.core.plugin.loader: QLibraryPrivate::loadPlugin failed on "/nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqxcb.so" : "Cannot load library /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/libqxcb.so: /nix/store/jq8mka96cq81mfzm4svzmbalna8asqfn-qtbase-6.10.1/lib/qt-6/plugins/platforms/../../../libQt6XcbQpa.so.6: undefined symbol: _ZTI28QBasicPlatformVulkanInstance, version Qt_6_PRIVATE_API"
qt.qpa.plugin: From 6.5.0, xcb-cursor0 or libxcb-cursor0 is needed to load the Qt xcb platform plugin.
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
qt.qpa.plugin: Could not find the Qt platform plugin "xcb,wayland" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, wayland-brcm, wayland-egl, wayland, xcb.

*************** Unhandled SIGABRT at 0x7f6a8d89df1c ***************
  0x0055dda1912588 LogCallstack [../build/../src/common/crash_handler.cpp:389]
  0x0055dda1912448 CrashSignalHandler [../build/../src/common/crash_handler.cpp:423]
  0x007f6a8d841d9f
  0x007f6a8d89df1c
  0x007f6a8d841ced
  0x007f6a8d828940
  0x007f6a8e0beda5
  0x007f6a8e0ff7c3
  0x007f6a8e0c0636
  0x007f6a8ed277f2
  0x007f6a8edd57a6
  0x007f6a8edd6217
  0x007f6a8e18b3f0
  0x007f6a8edd9214
  0x007f6a901985cc
  0x0055dda1602902 main [../build/../src/duckstation-qt/qthost.cpp:3573]
  0x007f6a8d82a4c4
  0x007f6a8d82a577
  0x0055dda1482754
  0xffffffffffffffff
*******************************************************************
Aborting application.
```
</details>

Apparently the bundled QT version tries to load plugins from the system KDE environment.

According to [Issue #238820 and it's comments](https://github.com/NixOS/nixpkgs/issues/238820#issuecomment-1599454411) there is a workaround by unsetting the `QT_PLUGIN_PATH=` environment variable. The issue also links to [it's corresponding discourse thread](https://discourse.nixos.org/t/appimage-run-fails-loading-qt-platform-plugin-xcb/29293/3). I also stumpled upon [PR #481707](https://github.com/NixOS/nixpkgs/pull/481707) which could resolve the issue in future, didn't try to update yet.

---

Enabling AppImage support via `binfmt` shortens the working command under KDE:
`QT_PLUGIN_PATH= ./DuckStation-x64.AppImage`

```Nix
programs = {
  appimage = {
    enable = true;
    binfmt = true;
  };
};
```

Otherwise don't forget to call it manually via `appimage-run`.
